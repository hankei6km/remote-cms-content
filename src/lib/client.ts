import { ClientBase, ClientKind, ClientOpts } from '../types/client.js'

export async function client(
  kind: ClientKind,
  opts: ClientOpts
): Promise<ClientBase> {
  if (kind === 'appsheet') {
    const { ClientAppSheet } = await import('./clients/appsheet.js')
    return new ClientAppSheet(opts)
  } else if (kind === 'contentful') {
    const { ClientCtf } = await import('./clients/contentful.js')
    return new ClientCtf(opts)
  } else if (kind === 'contentful:gql') {
    const { ClientCtfGql } = await import('./clients/contentful.js')
    return new ClientCtfGql(opts)
  } else if (kind === 'microcms') {
    const { ClientMicroCMS } = await import('./clients/microcms.js')
    return new ClientMicroCMS(opts)
  }
  throw new Error(`client: unknown kind ${kind}`)
}
