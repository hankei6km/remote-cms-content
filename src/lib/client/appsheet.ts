import axios from 'axios'
import {
  Client,
  ClientChain,
  ClientInstance,
  ClientOpts,
  FetchResult
} from '../../types/client.js'

export type APIActionBody = {
  Action: 'Find'
  Properties: any
  Rows: any[]
}

export function apiActionPath(
  appId: string,
  table: string,
  accessKey: string
): string {
  const q = new URLSearchParams()
  q.append('ApplicationAccessKey', accessKey)
  // https://api.appsheet.com/api/v2/ と連結することを想定.
  return `apps/${encodeURIComponent(appId)}/tables/${encodeURIComponent(
    table
  )}/Action?${q.toString()}`
}

export function apiActionBodyFind(
  props?: Record<string, string>
): APIActionBody {
  return {
    Action: 'Find',
    Properties: props || {},
    Rows: []
  }
}

export const client: Client = function client({
  apiBaseURL,
  apiName,
  credential
}: ClientOpts): ClientInstance {
  const request = () => {
    let skip: number | undefined = undefined
    let limit: number | undefined = undefined

    const clientChain: ClientChain = {
      limit(n: number) {
        limit = n
        return clientChain
      },
      skip(n: number) {
        skip = n
        return clientChain
      },
      async fetch(): Promise<FetchResult> {
        const res = await axios
          .post(
            `${apiBaseURL}${apiActionPath(
              credential[0],
              apiName || '',
              credential[1]
            )}`,
            JSON.stringify(apiActionBodyFind({})), // TODO: props の扱い.
            {
              headers: { 'Content-Type': ' application/json' }
            }
          )
          .catch((err) => {
            throw new Error(
              `client_appsheet.fetch API request error: table = ${apiName}, status = ${err.response.status}:${err.response.statusText}`
            )
          })
        if (res.data === '') {
          throw new Error(
            `client_appsheet.find API request error: table = ${apiName}, empty data received`
          )
        }

        return { rows: res.data }
      }
    }
    return clientChain
  }
  return { request }
}
