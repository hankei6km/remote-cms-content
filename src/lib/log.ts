import { Writable } from 'stream'
let _out: Writable | undefined
let _err: Writable | undefined

export function initLog(o: Writable | undefined, e: Writable | undefined) {
  _out = o
  _err = e
}

export function printInfo(s: string) {
  if (_out) {
    _out.write(s)
    _out.write('\n')
  }
}

export function printWarn(s: string) {
  if (_out) {
    _out.write(s)
    _out.write('\n')
  }
}

export function printErr(s: string) {
  if (_err) {
    _err.write(s)
    _err.write('\n')
  }
}
