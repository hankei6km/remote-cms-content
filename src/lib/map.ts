import path from 'path'
import { readFile } from 'fs/promises'
import Ajv from 'ajv'
import jsonata from 'jsonata'
import { BaseFlds, MapConfig, MapFld, MapFldsImage } from '../types/map.js'
import { mapConfigSchema } from '../types/mapConfigSchema.js'
import { ImageInfo } from '../types/media.js'
import { htmlToMarkdown } from './html.js'

const ajv = new Ajv()
const validate = ajv.compile(mapConfigSchema)

export function validateMapConfig(mapConfig: any): MapConfig {
  if (!validate(mapConfig) && validate.errors) {
    // allErrors false 前提.
    throw new Error(
      `validateMapConfig: message=${
        validate.errors[0]?.message
      } params=${JSON.stringify(validate.errors[0]?.params)}`
    )
  }
  return mapConfig as MapConfig
}

export async function loadMapConfig(jsonFile: string): Promise<MapConfig> {
  try {
    const s = await readFile(jsonFile)
    return validateMapConfig(JSON.parse(s.toString()))
  } catch (err) {
    throw new Error(`loadMapConfig: jsonFile=${jsonFile} ${err}`)
  }
}

export function fileNameFromURL(
  src: string,
  mapConfig: MapConfig,
  imageFld: MapFldsImage
): string {
  // client 側の処理にする？
  const fieldName =
    imageFld.fileNameField || mapConfig.media?.image?.fileNameField || ''
  let fileName = ''
  try {
    if (fieldName) {
      const q = new URLSearchParams(new URL(src).searchParams)
      fileName = path.basename(q.get(fieldName) || '')
    } else {
      const u = new URL(src)
      fileName = path.basename(u.pathname)
    }
  } catch (err: any) {
    throw new Error(
      `fileNameFromURL: src=${src},filedName=${fieldName}: ${err}`
    )
  }
  if (fileName === '') {
    throw new Error(
      `fileNameFromURL: src=${src},filedName=${fieldName}: image filename is blank`
    )
  }
  return fileName
}

export function isImageDownload(
  mapConfig: MapConfig,
  imageInfo: ImageInfo
): boolean {
  if (mapConfig.media?.image) {
    if (mapConfig.media.image.library) {
      const libIdx = mapConfig.media.image.library.findIndex(({ src }) =>
        imageInfo.url.startsWith(src)
      )
      if (libIdx >= 0) {
        return mapConfig.media.image.library[libIdx].download || false
      }
    }
    return mapConfig.media.image.download || false
  }
  return false
}

const validIdRegExp = /^[-_0-9a-zA-Z]+$/
export function validId(s: unknown): s is number | string {
  if (typeof s === 'number') {
    return true
  } else if (typeof s === 'string' && s.match(validIdRegExp)) {
    return true
  }
  return false
}

function throwInvalidId(
  value: unknown,
  srcName: string,
  dstName: string,
  fldType: string
) {
  throw new Error(
    `mappingFlds: invalid id: value = ${value}, params = ${srcName}, ${dstName}, ${fldType}`
  )
}

function throwInvalidType(
  actually: string,
  srcName: string,
  dstName: string,
  fldType: string
) {
  throw new Error(
    `mappingFlds: invalid type: actually type = ${actually}, params = ${srcName}, ${dstName}, ${fldType}`
  )
}

function transformFldValue(m: MapFld, value: unknown): any {
  const valueType = typeof value
  if (
    (valueType === 'number' ||
      valueType === 'string' ||
      valueType === 'object') &&
    m.jsonata
  ) {
    try {
      const expression = jsonata(m.jsonata)
      return expression.evaluate(value)
    } catch (err: any) {
      throw new Error(
        `transformFldValue: jsonata=${m.jsonata}, message=${err.message}`
      )
    }
  }
  return value
}

export async function mappingFlds(
  s: Record<string, unknown>,
  mapConfig: MapConfig
): Promise<BaseFlds> {
  const { _RowNumber, id, createdAt, updatedAt, ...src } = s
  const n = new Date()
  const ret: BaseFlds = {
    _RowNumber: typeof _RowNumber === 'number' ? _RowNumber! : -1,
    id: '',
    createdAt: typeof createdAt === 'string' ? new Date(createdAt) : n,
    updatedAt: typeof updatedAt === 'string' ? new Date(updatedAt) : n
  }
  if (validId(id)) {
    ret.id = `${id}`
  } else {
    throwInvalidId(id, 'id', 'id', 'id')
  }
  const mapFldsLen = mapConfig.flds.length
  for (let mapFldsIdx = 0; mapFldsIdx < mapFldsLen; mapFldsIdx++) {
    const m = mapConfig.flds[mapFldsIdx]
    if (src.hasOwnProperty(m.srcName)) {
      const srcValue = transformFldValue(m, src[m.srcName])
      const srcFldType = typeof srcValue
      switch (m.fldType) {
        case 'id':
          if (srcFldType === 'number' || srcFldType === 'string') {
            if (validId(srcValue)) {
              ret[m.dstName] = `${srcValue}`
            } else {
              throwInvalidId(srcValue, m.srcName, m.dstName, m.fldType)
            }
          } else {
            throwInvalidType(srcFldType, m.srcName, m.dstName, m.fldType)
          }
          break
        case 'boolean':
          if (srcFldType === 'boolean') {
            ret[m.dstName] = srcValue
          } else {
            throwInvalidType(srcFldType, m.srcName, m.dstName, m.fldType)
          }
          break
        case 'number':
          if (srcFldType === 'number') {
            ret[m.dstName] = srcValue
          } else {
            throwInvalidType(srcFldType, m.srcName, m.dstName, m.fldType)
          }
          break
        case 'string':
          if (srcFldType === 'string' || srcFldType === 'number') {
            ret[m.dstName] = `${srcValue}`
          } else {
            ret[m.dstName] = `${srcValue || ''}`
          }
          break
        case 'image': // この時点では文字列として扱う(保存時にファイルをダウンロードする).
          if (srcFldType === 'string' || srcFldType === 'number') {
            ret[m.dstName] = `${srcValue}`
          } else if (srcFldType === 'object') {
            ret[m.dstName] = srcValue
          } else {
            throwInvalidType(srcFldType, m.srcName, m.dstName, m.fldType)
          }
          break
        case 'datetime':
          ret[m.dstName] = new Date(`${srcValue}`)
          break
        case 'enum':
          const str = `${srcValue}`
          const matchIdx = m.replace.findIndex(({ pattern }) =>
            str.match(pattern)
          )
          if (matchIdx >= 0) {
            ret[m.dstName] = str.replace(
              m.replace[matchIdx].pattern,
              m.replace[matchIdx].replacement
            )
          } else {
            ret[m.dstName] = str
          }
          break
        case 'object':
          ret[m.dstName] = srcValue
          break
        case 'html':
          if (srcFldType === 'string') {
            ret[m.dstName] = await htmlToMarkdown(srcValue as string, {
              embedImgAttrs: m.embedImgAttrs
            })
          } else {
            throwInvalidType(srcFldType, m.srcName, m.dstName, m.fldType)
          }
          break
      }
      if (ret.hasOwnProperty(m.dstName)) {
        delete src[m.srcName]
      }
    }
  }
  if (mapConfig.passthruUnmapped) {
    Object.entries(src).forEach(([k, v]) => {
      if (!ret.hasOwnProperty(k)) {
        ret[k] = v
      }
    })
  }
  return ret
}
