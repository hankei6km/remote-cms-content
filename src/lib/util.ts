import { readFileSync } from 'fs'

export function readQuery(
  fileName: string,
  validator?: (text: string) => Error | undefined
): string {
  try {
    const t = readFileSync(fileName).toString()
    if (typeof validator === 'function') {
      const err = validator(t)
      if (err) {
        throw new Error(`validation: err=${err}`)
      }
    }
    return t
  } catch (err: any) {
    throw new Error(`readQuery: fileName=${fileName} err=${err}`)
  }
}

export function apolloErrToMessages(err: any): string {
  if (
    typeof err === 'object' &&
    typeof err.message === 'string' &&
    err.graphQLErrors
  ) {
    return [
      err.message,
      ...Object.entries(err)
        .filter(
          ([_k, v]: [string, any]) => v && Array.isArray(v.result?.errors)
        )
        .map(([k, v]: [string, any]) =>
          v.result.errors
            .map(({ message }: any) => `${k}: ${message}`)
            .join('\n')
        )
    ].join('\n')
  }
  return err
}
