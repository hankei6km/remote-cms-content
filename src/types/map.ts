import { Expression } from 'jsonata'
import imageSalt from '@hankei6km/rehype-image-salt'

export type BaseFlds = {
  id: string
  _RowNumber: number
  createdAt: Date
  updatedAt: Date
}

export type MappedFlds = {
  id: string
} & Record<string, unknown>

export type MapFldsBase = {
  fetchFld?: string
  query: string
  dstName: string
  /**
   * @ignore
   */
  transformJsonata?: Expression
}

export type MapFldsId = {
  fldType: 'id'
} & MapFldsBase

export type MapFldsBoolean = {
  fldType: 'boolean'
} & MapFldsBase

export type MapFldsNumber = {
  fldType: 'number'
} & MapFldsBase

export type MapFldsString = {
  fldType: 'string'
} & MapFldsBase

export type MapFldsDatetime = {
  fldType: 'datetime'
} & MapFldsBase

export type MapFldsImage = {
  fldType: 'image'
  fileNameField?: string
  setSize?: boolean
} & MapFldsBase

export type MapFldsEnum = {
  fldType: 'enum'
  replace: {
    pattern: string | RegExp
    replacement: string
  }[]
} & MapFldsBase

export type MapFldsObject = {
  fldType: 'object' // 実質 any
} & MapFldsBase

export type HtmlToHtmlOpts = {
  frontMatter?: boolean
  splitParagraph?: boolean
  lfTo?: string
}

export type HtmlToMarkdownOpts = {
  imageSalt?: Parameters<typeof imageSalt>[0]
}
export type MapFldsHtmlOpts = {
  convert?: 'none' | 'html' | 'markdown'
  toHtmlOpts?: HtmlToHtmlOpts
  toMarkdownOpts?: HtmlToMarkdownOpts
}
export type MapFldsHtml = {
  fldType: 'html'
  convert?: MapFldsHtmlOpts['convert']
  toHtmlOpts?: MapFldsHtmlOpts['toHtmlOpts']
  toMarkdownOpts?: MapFldsHtmlOpts['toMarkdownOpts']
} & MapFldsBase

export type MapFld =
  | MapFldsId
  | MapFldsBoolean
  | MapFldsNumber
  | MapFldsString
  | MapFldsDatetime
  | MapFldsImage
  | MapFldsEnum
  | MapFldsObject
  | MapFldsHtml

export type MapFlds = MapFld[]

export type MapConfig = {
  media?: {
    image?: {
      fileNameField?: string
      download?: boolean
      library?: [
        {
          src: string
          kind: 'imgix'
          download?: boolean
        }
      ]
    }
  }
  disableBaseFlds?: boolean
  passthruUnmapped?: boolean
  selectFldsToFetch?: boolean
  fldsToFetch?: string[]
  transform?: string
  /**
   * @ignore
   */
  transformJsonata?: Expression
  position?: {
    disable?: boolean
    fldName?: string
    start?: number
  }
  flds: MapFlds
}

export const defaultPosition = {
  disable: false,
  fldName: 'position',
  start: 1
}
