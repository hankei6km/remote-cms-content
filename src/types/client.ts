import { MapFld } from './map.js'

export const ClientKindValues = ['appsheet', 'contentful', 'microcms'] as const
export type ClientKind = typeof ClientKindValues[number]

// type GetFldSyncType = Exclude<MapFld, MapFldsImage | MapFldsHtml>
// type GetFldAsyncType = Exclude<MapFld, GetFldSyncType>
export class ResRecord {
  record: Record<string, unknown> = {}
  _RowNumber: unknown
  id: unknown
  createdAt: unknown
  updatedAt: unknown
  src: Record<string, unknown> = {}
  constructor(record: Record<string, unknown>) {
    const { _RowNumber, id, createdAt, updatedAt, ...r } = record
    this.record = record
    this.src = r
    this._RowNumber = _RowNumber
    this.id = id
    this.createdAt = createdAt
    this.updatedAt = updatedAt
  }
  has(map: MapFld): boolean {
    return this.record.hasOwnProperty(map.srcName)
  }
  baseFlds() {
    return {
      _RowNumber: this._RowNumber,
      id: this.id,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt
    }
  }
  rawEntries(): [string, unknown][] {
    return Object.entries(this.src) // ここでは get* メソッドを通さない.
  }
  isAsyncFld(map: MapFld): boolean {
    // return map.fldType === 'html' || map.fldType === 'image'
    return false
  }
  getSync(map: MapFld): unknown {
    return this.record[map.srcName]
  }
  async getAsync(map: MapFld): Promise<unknown> {
    return this.record[map.srcName]
  }
}

export type FetchResult = {
  content: ResRecord[]
}

export const OpKindValues = ['eq'] as const
export type OpKind = typeof OpKindValues[number]
export type OpValue = [OpKind, string, any]

export type TransformContent = (
  content: Record<string, unknown>[]
) => Record<string, unknown>[]

export type ClientChain = {
  api: (name: string) => ClientChain
  filter: (o: OpValue[]) => ClientChain
  limit: (n: number) => ClientChain
  skip: (n: number) => ClientChain
  pageSize(n: number): ClientChain
  transform: (t: TransformContent) => ClientChain
  fetch: () => Promise<FetchResult>
}

export type ClientOpts = {
  apiBaseURL: string
  apiName?: string
  credential: string[]
}

export abstract class ClientBase {
  _opts: ClientOpts
  _apiName: string = ''
  _filter: OpValue[] = []
  _skip: number | undefined = undefined
  _limit: number | undefined = undefined
  _pageSize: number | undefined = undefined
  _transformer: TransformContent | undefined = undefined

  constructor(opts: ClientOpts) {
    this._opts = opts
    this._apiName = opts.apiName || ''
  }

  api(name: string): ClientChain {
    this._apiName = name
    return this
  }
  filter(o: OpValue[]): ClientChain {
    this._filter.push(...o)
    return this
  }
  limit(n: number): ClientChain {
    this._limit = n
    return this
  }
  skip(n: number): ClientChain {
    this._skip = n
    return this
  }
  pageSize(n: number): ClientChain {
    this._pageSize = n
    return this
  }
  transform(t: TransformContent): ClientChain {
    this._transformer = t
    return this
  }
  abstract fetch(): Promise<FetchResult>
  // clientChain: ClientChain = {
  //   api: this.api.bind(this),
  //   filter: this.filter.bind(this),
  //   limit: this.limit.bind(this),
  //   skip: this.skip.bind(this),
  //   transform: this.transform.bind(this),
  //   fetch: this.fetch.bind(this)
  // }

  abstract kind(): ClientKind // なくてもいいかな.
  request(): ClientChain {
    return this
  }
}

export type Client = (opst: ClientOpts) => ClientBase
