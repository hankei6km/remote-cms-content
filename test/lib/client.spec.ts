import { jest } from '@jest/globals'
import { initLog } from '../../src/lib/log.js'
import { compileMapFld } from '../../src/lib/map.js'
import { ResRecord } from '../../src/types/client.js'
import { mockStreams } from '../util.js'

// > ENOENT: no such file or directory, open 'zlib'
// になる対応.
// contentful を import すると発生するが原理は不明.
jest.unstable_mockModule('contentful', async () => {
  return {
    default: jest.fn()
  }
})

const { readQuery: _readQuery, ...utils } = await import(
  '../../src/lib/util.js'
)
jest.unstable_mockModule('../../src/lib/util.js', async () => {
  return {
    readQuery: (v: string) => (v === 'error' ? new Error('query error') : v), // ファイル読み込みではなく、文字列をそのまま返す.
    ...utils
  }
})

afterEach(() => {
  initLog(undefined, undefined)
})

const { client } = await import('../../src/lib/client.js')
const { ClientTest } = await import('./clientTest.js')

describe('ResRecord', () => {
  it('should return false from isAsyncFld', () => {
    expect(
      new ResRecord({}).isAsyncFld(
        compileMapFld({
          query: '',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toBeFalsy()
    expect(
      new ResRecord({}).isAsyncFld(
        compileMapFld({
          query: '',
          dstName: '',
          fldType: 'html'
        })
      )
    ).toBeFalsy()
  })
  it('should get the value of field', () => {
    expect(
      new ResRecord({ text: 'test1' }).getSync(
        compileMapFld({
          query: 'text',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toEqual('test1')
  })
  it('should get the value of field that is contained "fields"', () => {
    expect(
      new ResRecord({ fields: { text: 'test1' } }).getSync(
        compileMapFld({
          query: 'fields.text',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toEqual('test1')
  })
  it('should return undefined', () => {
    expect(
      new ResRecord({ text: 'test1' }).getSync(
        compileMapFld({
          query: 'fields.abc',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toEqual(undefined)
    expect(
      new ResRecord({ fields: { text: 'test1' } }).getSync(
        compileMapFld({
          query: 'fields.abc',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toEqual(undefined)
    expect(
      new ResRecord({ text: 'test1' }).getSync(
        compileMapFld({
          query: 'fields.text',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toEqual(undefined)
  })
  it('should return undefined', () => {
    expect(
      new ResRecord({ text: 'test1' }).getSync(
        compileMapFld({
          query: 'abc',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toBeUndefined()
    expect(
      new ResRecord({ text: 'test1' }).getSync(
        compileMapFld({
          query: 'fields.text',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toBeUndefined()
    expect(
      new ResRecord({ fields: { text: 'test1' } }).getSync(
        compileMapFld({
          query: 'fields.abc',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toBeUndefined()
  })
  it('should return null', () => {
    expect(
      new ResRecord({ text: null }).getSync(
        compileMapFld({
          query: 'text',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toBeNull()
    expect(
      new ResRecord({ fields: { text: null } }).getSync(
        compileMapFld({
          query: 'fields.text',
          dstName: '',
          fldType: 'string'
        })
      )
    ).toBeNull()
  })
})

describe('ClientBase', () => {
  it('should fetch all content', async () => {
    const c = new ClientTest({ apiBaseURL: '', credential: [] }).genRecord(100)
    const g = c.fetch()
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 100 },
        content: c._record.map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 0,
      pageSize: undefined,
      endCursor: null,
      query: []
    })
    expect(await g.next()).toEqual({
      value: undefined,
      done: true
    })
    expect(c._fetch).toHaveBeenCalledTimes(1)
  })
  it('should fetch all content with paginate', async () => {
    const c = new ClientTest({ apiBaseURL: '', credential: [] }).genRecord(100)
    const g = c.pageSize(30).fetch()
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 30 },
        content: c._record.slice(0, 30).map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 0,
      pageSize: 30,
      endCursor: null,
      query: []
    })
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 30 },
        content: c._record.slice(30, 60).map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 30,
      pageSize: 30,
      endCursor: null,
      query: []
    })
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 30 },
        content: c._record.slice(60, 90).map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 60,
      pageSize: 30,
      endCursor: null,
      query: []
    })
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 10 },
        content: c._record.slice(90, 100).map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 90,
      pageSize: 30,
      endCursor: null,
      query: []
    }) // limit を指定していない.
    expect(await g.next()).toEqual({
      value: undefined,
      done: true
    })
    expect(c._fetch).toHaveBeenCalledTimes(4)
  })
  it('should fetch part of content', async () => {
    const c = new ClientTest({ apiBaseURL: '', credential: [] }).genRecord(100)
    const g = c.skip(5).limit(11).fetch()
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 11 },
        content: c._record.slice(5, 16).map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 5,
      pageSize: undefined,
      endCursor: null,
      query: []
    })
    expect(c._fetch).toHaveBeenCalledTimes(1)
  })
  it('should fetch part of content with paginate', async () => {
    const c = new ClientTest({ apiBaseURL: '', credential: [] }).genRecord(100)
    const g = c.skip(5).pageSize(30).limit(75).fetch()
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 30 },
        content: c._record.slice(5, 35).map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 5,
      pageSize: 30,
      endCursor: null,
      query: []
    })
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 30 },
        content: c._record.slice(35, 65).map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 35,
      pageSize: 30,
      endCursor: null,
      query: []
    })
    expect(await g.next()).toEqual({
      value: {
        fetch: { next: { kind: 'total', total: 100 }, count: 15 },
        content: c._record.slice(65, 80).map((v) => new ResRecord(v))
      },
      done: false
    })
    expect(c._fetch).toHaveBeenLastCalledWith({
      skip: 65,
      pageSize: 15,
      endCursor: null,
      query: []
    })
    expect(await g.next()).toEqual({
      value: undefined,
      done: true
    })
    expect(c._fetch).toHaveBeenCalledTimes(3)
  })
  it('should print info from fetch method', async () => {
    const info = { o: '', e: '' }
    initLog(...mockStreams(info))
    const c = new ClientTest({ apiBaseURL: '', credential: [] }).genRecord(100)
    for await (let _res of c.pageSize(30).fetch()) {
    }
    expect(info.o).toMatchSnapshot()
    expect(info.e).toEqual('')
  })
  it('should print info from fetch method(limit)', async () => {
    const info = { o: '', e: '' }
    initLog(...mockStreams(info))
    const c = new ClientTest({ apiBaseURL: '', credential: [] }).genRecord(100)
    for await (let _res of c.pageSize(30).limit(50).fetch()) {
    }
    expect(info.o).toMatchSnapshot()
    expect(info.e).toEqual('')
  })
  it('should throw error when error occured in chain(query)', async () => {
    const c = new ClientTest({ apiBaseURL: '', credential: [] }).query([
      'error'
    ])
    const g = c.fetch()
    await expect(g.next()).rejects.toThrowError(/query/)
  })
  it('should throw error when error occured in chain(vars)', async () => {
    const c = new ClientTest({ apiBaseURL: '', credential: [] }).vars(
      ['error'],
      true
    )
    const g = c.fetch()
    await expect(g.next()).rejects.toThrowError(/vars/)
  })
})

// https://github.com/facebook/jest/issues/11438
// 実際の client を使うのはここでのみ.
// 1 つの worker を共有する複数の spec の import 先で動的 import を使うとエラーになる.
// その対策.
describe('client', () => {
  it('should return appsheet client instanse', async () => {
    const c = await client('appsheet', {
      apiBaseURL: 'http://localhost:3000',
      credential: ['id', 'secret']
    })
    expect(c.kind()).toEqual('appsheet')
  })
  it('should return contentful client instanse', async () => {
    const c = await client('contentful', {
      apiBaseURL: '',
      credential: ['spaceId', 'cda_token']
    })
    expect(c.kind()).toEqual('contentful')
  })
  it('should return contentful:gql client instanse', async () => {
    const c = await client('contentful:gql', {
      apiBaseURL: 'https://graphql.contentful.com/content/v1/spaces/',
      credential: ['spaceId', 'cda_token']
    })
    expect(c.kind()).toEqual('contentful:gql')
  })
  it('should return graphcms:gql client instanse', async () => {
    const c = await client('graphcms:gql', {
      apiBaseURL: 'https://[region].graphcms.com/v2/',
      credential: ['projectId', 'pat'],
      apiName: 'master'
    })
    expect(c.kind()).toEqual('graphcms:gql')
  })
  it('should return prismic:gql client instanse', async () => {
    const c = await client('prismic:gql', {
      apiBaseURL: '',
      credential: ['repository-name', 'pat']
    })
    expect(c.kind()).toEqual('prismic:gql')
  })
  it('should return microcms client instanse', async () => {
    const c = await client('microcms', {
      apiBaseURL: 'http://localhost:3000',
      credential: ['X-MICROCMS-API-KEY', 'secret']
    })
    expect(c.kind()).toEqual('microcms')
  })
  it('should throw error when pass unkown kid', async () => {
    await expect(() =>
      client('UNKWON' as any, {
        apiBaseURL: 'http://localhost:3000',
        credential: ['id', 'secret']
      })
    ).rejects.toThrowError('client: unknown kind UNKWON')
  })
})

export {}
