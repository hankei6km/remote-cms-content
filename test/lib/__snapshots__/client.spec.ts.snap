// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClientBase should print info from fetch method 1`] = `
"ClientBase.fetch: start
ClientBase.fetch: repeat=1 skip=0, pageSize=30
ClientBase.fetch:   count=30
ClientBase.fetch: repeat=2 skip=30, pageSize=30
ClientBase.fetch:   count=30
ClientBase.fetch: repeat=3 skip=60, pageSize=30
ClientBase.fetch:   count=30
ClientBase.fetch: repeat=4 skip=90, pageSize=30
ClientBase.fetch:   count=10
"
`;

exports[`ClientBase should print info from fetch method(limit) 1`] = `
"ClientBase.fetch: start limit=50
ClientBase.fetch: repeat=1 skip=0, pageSize=30
ClientBase.fetch:   count=30
ClientBase.fetch: repeat=2 skip=30, pageSize=20
ClientBase.fetch:   count=20
"
`;
